File: ui/src/review.js
/* Reads what dagger says about a change and works out where to draw it.
 *
 * Nothing here decides anything about the code. Which definitions are worth reading, what
 * order to read them in, what broke what — all of that arrives already settled, and this
 * only places it on a page. Where the mock computed a reading order in the browser, it now
 * comes from the review itself.
 *
 * Plain functions, no components: this is the part worth testing without a browser.
 */

export const NODE_H = 28;
const ROW_GAP = 22, GROUP_GAP = 30, FILE_GAP = 14;
const PAD_X = 12, PAD_TOP = 24, PAD_BOTTOM = 10, NODE_GAP = 10, MARGIN = 16;

export const MARK = {
  added: "+", removed: "−", contract: "!", body: "~", docs: '"', affected: "≈", still: "·",
};
export const TINT = {
  added: "add", removed: "del", contract: "chg", body: "chg", docs: "chg", affected: "aff", still: "aff",
};
/* Louder first, so a row of nodes reads worst-first when nothing else decides the order. */
const LOUDNESS = ["contract", "removed", "added", "body", "docs", "affected", "still"];

/* How wide a character is in the graph's font, which has to agree with style.css: a
 * monospace advance is 0.6 of its size, and the boxes are drawn from this rather than
 * measured, since there's no page to measure against when this is tested. */
const FONT = 14;
const CHAR = FONT * 0.6;
/* Past this a name is cut short. A test is a sentence with underscores in it, and one of
 * them is as wide as a dozen ordinary definitions put together — which buries them. */
const LONGEST = 22;

/* A box wears its name along the top, so it can't be narrower than the name. Smaller than
 * the nodes' font, and it has to agree with style.css the same way. */
const BOX_FONT = 12.5;
const labelWidth = (text) => Math.ceil(text.length * BOX_FONT * 0.6) + 22;

/** A name as the graph shows it: long ones lose their tail rather than their box. */
export const shorten = (name) =>
  name.length > LONGEST ? `${name.slice(0, LONGEST - 1)}…` : name;

/* Rounded up, never down: a box half a pixel too small is a name poking out of it. The 34
 * is the space either side plus the mark and the gap after it. */
export const widthOf = (text) => Math.ceil(shorten(text).length * CHAR) + 34;

/* What to call a module whose own definition isn't in this review, so its name never got
 * reported. The file it lives in is the best guess left, minus the extension — which is
 * about the file on disk, not about the code. */
const guessed = (path) => path.split("/").pop().replace(/\.[^.]+$/, "");

/* ---------------- what dagger said, in the shapes a page wants ---------------- */

export function digest(raw) {
  const definitions = new Map();
  for (const definition of raw.definitions) {
    const sides = definition.sides;
    const before = sides.kept ? sides.kept.before : sides.removed || null;
    const after = sides.kept ? sides.kept.after : sides.added || null;
    const shown = after || before;

    definitions.set(definition.identity, {
      id: definition.identity,
      name: shown.locator.name,
      path: [...shown.locator.scope, shown.locator.name].join("::"),
      file: shown.file,
      kind: shown.kind,
      before,
      after,
      mark: marking(raw.review, definition.identity),
      group: (raw.grouping.of || {})[definition.identity] || [],
    });
  }

  return {
    definitions,
    steps: raw.ordering.steps.filter((step) => definitions.has(step.definition)),
    edges: raw.review.edges.filter((e) => definitions.has(e.from) && definitions.has(e.to)),
    changes: raw.review.changes,
    cost: raw.ordering.cost,
    grouping: raw.grouping.name,
    worries: [
      ...raw.review.diagnostics.map(told),
      ...raw.notes.map((note) => (note.file ? `${note.file}: ${note.message}` : note.message)),
    ],
  };
}

/* One word for what happened, which is all a node has room for. */
function marking(review, id) {
  const change = review.changes[id];
  if (change === "added") return "added";
  if (change === "removed") return "removed";

  const edits = change.kept;
  if (edits.contract) return "contract";
  if (edits.parts.includes("body") || edits.parts.includes("type")) return "body";
  if (edits.parts.includes("docs")) return "docs";
  return review.affected.includes(id) ? "affected" : "still";
}

/** Whether a change to this one means its callers have to change too. */
export function broke(review, id) {
  const change = review.changes[id];
  if (change === "removed") return true;
  if (change === "added") return false;
  return Boolean(change.kept && change.kept.contract);
}

/* Dagger's diagnostics say what it had to work around. A reader deserves them in words
 * rather than as a shape of JSON. */
function told(diagnostic) {
  const [[kind, what]] = Object.entries(diagnostic);
  switch (kind) {
    case "unattributed":
      return `${what.file}: ${what.lines} changed line${what.lines === 1 ? "" : "s"} belong to no definition, around line ${what.at.join(", ")}`;
    case "lopsided_contract":
      return "one side of a definition had a compiler's word on it and the other didn't, so the signature as written was compared instead";
    case "unbound_in_contract":
      return `${what.symbol} appears where callers can see it, but nothing could say what it refers to`;
    case "mention_from_nowhere":
      return `a mention came from ${what.from.name}, which was never reported`;
    default:
      return kind;
  }
}

/* ---------------- laying it out ---------------- */

/* Boxes inside boxes: a group holds files, a file holds definitions. Whatever holds
 * something else up is drawn above it, so the page reads downwards the way the review does.
 *
 * A module is its file, so it gives the file's box its name rather than taking a node inside
 * the box that stands for the box.
 */
export function layout(review) {
  const nodes = [...review.definitions.values()].filter((d) => d.kind !== "module");
  const modules = new Map();
  for (const definition of review.definitions.values()) {
    if (definition.kind === "module") modules.set(definition.file, definition);
  }

  const leansOn = new Map(nodes.map((n) => [n.id, []]));
  for (const edge of review.edges) {
    if (leansOn.has(edge.from) && leansOn.has(edge.to)) leansOn.get(edge.from).push(edge.to);
  }

  const byFile = collect(nodes, (node) => node.file);
  const byGroup = collect([...byFile.keys()], (file) => byFile.get(file)[0].group.join("/"));
  const laning = stacking(byFile, review.edges);

  const boxes = [...byGroup].map(([groupPath, filePaths]) => {
    const stacked = laning(filePaths);
    const files = [...filePaths].sort((a, b) => stacked(a) - stacked(b)).map((path) => {
      const rows = layer(byFile.get(path), leansOn);
      const inner = Math.max(...rows.map(rowWidth));
      const module = modules.get(path) || null;
      const label = module ? module.name : guessed(path);
      return {
        key: path,
        module,
        label,
        rows,
        /* A module's box also carries its mark, so there are two more characters to fit. */
        w: Math.max(inner + 2 * PAD_X, labelWidth(module ? `${label}xx` : label)),
        h: rows.length * NODE_H + (rows.length - 1) * ROW_GAP + PAD_TOP + PAD_BOTTOM,
      };
    });

    /* Files that hold each other up stack; files that have nothing to do with each other
     * sit side by side. Stacking those anyway is what made a group a single tall column
     * with the page empty either side of it. */
    const lanes = [];
    for (const file of files) (lanes[stacked(file.key)] ||= []).push(file);
    const inLanes = lanes.filter(Boolean);

    return {
      key: groupPath,
      label: groupPath || "elsewhere",
      files,
      lanes: inLanes,
      w: Math.max(
        Math.max(...inLanes.map(laneWidth)) + 2 * PAD_X,
        labelWidth(groupPath || "elsewhere"),
      ),
      h:
        inLanes.reduce((sum, lane) => sum + laneHeight(lane), 0) +
        FILE_GAP * (inLanes.length - 1) +
        PAD_TOP +
        PAD_BOTTOM,
    };
  });

  const at = new Map();
  let y = MARGIN;
  let widest = 0;

  for (const band of bands(boxes, review)) {
    let x = MARGIN;
    for (const box of band) {
      box.x = x;
      box.y = y;
      let fileY = y + PAD_TOP;

      for (const lane of box.lanes) {
        let fileX = x + PAD_X;

        for (const file of lane) {
          file.x = fileX;
          file.y = fileY;
          let rowY = fileY + PAD_TOP;

          for (const row of file.rows) {
            let nodeX = file.x + PAD_X + (file.w - 2 * PAD_X - rowWidth(row)) / 2;
            for (const node of row) {
              at.set(node.id, { x: nodeX, y: rowY, w: widthOf(node.name) });
              nodeX += widthOf(node.name) + NODE_GAP;
            }
            rowY += NODE_H + ROW_GAP;
          }
          fileX += file.w + FILE_GAP;
        }
        fileY += laneHeight(lane) + FILE_GAP;
      }
      x += box.w + GROUP_GAP;
    }
    widest = Math.max(widest, x - GROUP_GAP + MARGIN);
    y += Math.max(...band.map((box) => box.h)) + GROUP_GAP;
  }

  return { at, boxes, modules, w: widest, h: y - GROUP_GAP + MARGIN };
}

/* Rows within a box: something sits below everything it leans on. */
function layer(nodes, leansOn) {
  const here = new Set(nodes.map((n) => n.id));
  const rows = [];
  for (const node of nodes) {
    const row = depth(node.id, here, leansOn, new Map());
    (rows[row] ||= []).push(node);
  }
  for (const row of rows) if (row) row.sort(byLoudness);
  /* Row nought is whatever leans on nothing, and it goes at the top: a reader meets what
   * holds things up before the things it holds. */
  return rows.filter(Boolean).flatMap(folded);
}

/* A row of peers folded into a block about as wide as it is tall, so a file with a dozen
 * tests in it grows downwards instead of off the side of the page. Nothing in a row leans
 * on anything else in it — an edge would have put one of them a row lower — so they can be
 * split across lines without a line ever pointing the wrong way. */
function folded(row) {
  const across = Math.ceil(Math.sqrt(row.length));
  const lines = [];
  for (let at = 0; at < row.length; at += across) lines.push(row.slice(at, at + across));
  return lines;
}

/* Files stack in a group the same way nodes stack in a file and groups stack on the page:
 * whatever holds another file up sits above it. Without this the files in a group land in
 * whatever order they turned up in, and half the lines between them run the wrong way. */
function stacking(byFile, edges) {
  const fileOf = new Map();
  for (const [path, inside] of byFile) for (const node of inside) fileOf.set(node.id, path);

  const leansOn = new Map([...byFile.keys()].map((path) => [path, []]));
  for (const edge of edges) {
    const from = fileOf.get(edge.from);
    const to = fileOf.get(edge.to);
    if (from && to && from !== to) leansOn.get(from).push(to);
  }

  /* Depth is worked out one group at a time, counting only what that group holds. A file
   * leaning on something in another group says nothing about where it belongs in this one —
   * which lane it lands in is a question about its neighbours — and letting those outside
   * edges count pushed files below others that weren't holding them up at all. Where the
   * groups themselves go is bands()' job. */
  return (paths) => {
    const within = new Set(paths);
    const seen = new Map();
    return (path) => depth(path, within, leansOn, seen);
  };
}

/* Bands of boxes: whatever holds another box up is drawn in an earlier band. */
function bands(boxes, review) {
  const groupOf = new Map();
  for (const box of boxes) {
    for (const file of box.files) {
      for (const row of file.rows) for (const node of row) groupOf.set(node.id, box.key);
    }
  }

  const leansOn = new Map(boxes.map((box) => [box.key, []]));
  for (const edge of review.edges) {
    const from = groupOf.get(edge.from);
    const to = groupOf.get(edge.to);
    if (from && to && from !== to) leansOn.get(from).push(to);
  }

  const keys = new Set(boxes.map((box) => box.key));
  const found = [];
  for (const box of boxes) {
    const row = depth(box.key, keys, leansOn, new Map());
    (found[row] ||= []).push(box);
  }
  return found.filter(Boolean);
}

/* How far above the bottom something sits: one more than the furthest thing it leans on.
 * A circle is settled by whoever is asked first, which is enough — being in a circle means
 * there is no right answer, only a readable one. */
function depth(id, within, leansOn, seen) {
  if (seen.has(id)) return seen.get(id);
  seen.set(id, 0);
  const below = (leansOn.get(id) || []).filter((other) => within.has(other) && other !== id);
  const found = below.length
    ? 1 + Math.max(...below.map((other) => depth(other, within, leansOn, seen)))
    : 0;
  seen.set(id, found);
  return found;
}

const byLoudness = (a, b) =>
  LOUDNESS.indexOf(a.mark) - LOUDNESS.indexOf(b.mark) || a.name.localeCompare(b.name);
const rowWidth = (row) => row.reduce((sum, n) => sum + widthOf(n.name), 0) + NODE_GAP * (row.length - 1);
const laneWidth = (lane) => lane.reduce((sum, f) => sum + f.w, 0) + FILE_GAP * (lane.length - 1);
const laneHeight = (lane) => Math.max(...lane.map((f) => f.h));

function collect(items, by) {
  const out = new Map();
  for (const item of items) {
    const key = by(item);
    if (!out.has(key)) out.set(key, []);
    out.get(key).push(item);
  }
  return out;
}

/* ---------------- the text of a definition ---------------- */

/* One diff, not one per part. The split into parts decides what breaks callers; it isn't how
 * anybody reads code. Where a part's pieces aren't next to each other in the file — a
 * module's imports, an implementation's braces — a gap stands in rather than pretending the
 * lines met. */
export function stitch(occurrence) {
  if (!occurrence) return null;

  const pieces = Object.values(occurrence.parts)
    .flat()
    .sort((a, b) => a.span.start - b.span.start);

  let out = "";
  let last = null;
  for (const piece of pieces) {
    /* Pieces that touch are run together exactly as the file has them. Putting a newline
     * between them instead is how a signature and its opening brace ended up on separate
     * lines. Only a real gap gets a line of its own. */
    if (last !== null && piece.span.start > last) out += "\n…\n";
    out += piece.text;
    last = piece.span.end;
  }

  return straighten(out.replace(/^\n+|\n+$/g, ""));
}

/* A definition starts at its name rather than at the margin, so its first line turns up
 * without the indentation every line beneath it still carries. Taking that much off the
 * rest lines them up the way the file has them. */
function straighten(text) {
  const lines = text.split("\n");
  const under = lines.slice(1).filter((line) => line.trim() && line !== "…");
  if (!under.length) return text;

  const spare = Math.min(...under.map((line) => line.match(/^ */)[0].length));
  if (!spare) return text;

  return [lines[0], ...lines.slice(1).map((line) => line.slice(spare))].join("\n");
}

/* Names in this review that can be pointed at without ambiguity.
 *
 * What a highlighter can't know: which words in this code are things the reader is about to
 * read, or has just read. A name appearing twice is left alone — pointing at the wrong one
 * is worse than pointing at nothing. */
export function namesIn(review) {
  const seen = new Map();
  for (const definition of review.definitions.values()) {
    seen.set(definition.name, seen.has(definition.name) ? null : definition.id);
  }
  for (const [name, id] of seen) if (id === null) seen.delete(name);
  return seen;
}

const TOKENS = /(\/\/[^\n]*|#[^\n]*|\/\*[\s\S]*?(?:\*\/|$))|("(?:[^"\\]|\\.)*"?|'(?:[^'\\]|\\.)*'?|`(?:[^`\\]|\\.)*`?)|([A-Za-z_$][\w$]*)|([\s\S])/g;

/** One line of code, split into what it's made of. */
export function tokens(line) {
  const out = [];
  let match;
  TOKENS.lastIndex = 0;

  while ((match = TOKENS.exec(line))) {
    const [, comment, string, name, other] = match;
    const kind = comment ? "quiet" : string ? "quiet" : name ? "name" : "plain";
    const text = comment || string || name || other;

    const last = out[out.length - 1];
    if (last && last.kind === kind && kind === "plain") last.text += text;
    else out.push({ kind, text });
  }
  return out;
}

/** Line by line, marked as kept, gone, or new. */
export function compare(before, after) {
  const a = before === null ? [] : before.split("\n");
  const b = after === null ? [] : after.split("\n");
  const same = Array.from({ length: a.length + 1 }, () => new Array(b.length + 1).fill(0));

  for (let i = a.length - 1; i >= 0; i--) {
    for (let j = b.length - 1; j >= 0; j--) {
      same[i][j] = a[i] === b[j] ? same[i + 1][j + 1] + 1 : Math.max(same[i + 1][j], same[i][j + 1]);
    }
  }

  const out = [];
  let i = 0;
  let j = 0;
  while (i < a.length && j < b.length) {
    if (a[i] === b[j]) { out.push([" ", a[i]]); i++; j++; }
    else if (same[i + 1][j] >= same[i][j + 1]) out.push(["−", a[i++]]);
    else out.push(["+", b[j++]]);
  }
  while (i < a.length) out.push(["−", a[i++]]);
  while (j < b.length) out.push(["+", b[j++]]);
  return out;
}

File: crates/core/src/review.rs
//! Everything one review of a repository comes to.
//!
//! One object, built in one place at the end, with everything about a definition on that
//! definition, so nothing can name a definition that wasn't handed over.

use crate::change::{Change, Mark, classify};
use crate::diagnostic::Diagnostic;
use crate::group::Grouping;
use crate::model::{self, Identity, Locator, Role, Sides};
use crate::order::{Cost, Step, order};
use crate::propagate::reached;
use crate::reference::{Reference, Target};
use crate::shape::{Group, Shape};
use serde::{Deserialize, Serialize};
use std::collections::{BTreeMap, BTreeSet, VecDeque};
use std::num::NonZeroU32;

/// One definition depending on another, as the reader will see it. Definitions that
/// nobody needs to read are left out, so an edge can stand in for a chain that ran
/// through them.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[cfg_attr(feature = "typescript", derive(ts_rs::TS))]
pub struct Edge {
    pub from: Identity,
    pub to: Identity,
}

/// A dependency between two of `places`' members, as their places. Nothing when either
/// end isn't among them, or both ends are the same one.
pub(crate) fn placed(
    places: &BTreeMap<Identity, usize>,
    from: Identity,
    to: Identity,
) -> Option<(usize, usize)> {
    places
        .get(&from)
        .zip(places.get(&to))
        .map(|(&from, &to)| (from, to))
        .filter(|(from, to)| from != to)
}

/// How much trust a warning costs the review it's about.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[cfg_attr(feature = "typescript", derive(ts_rs::TS))]
#[serde(rename_all = "snake_case")]
pub enum Impact {
    /// Something that changed might not be here at all, and a reader trusting the review
    /// would never learn of it.
    Incomplete,
    /// It's here and it's shown, but it was worked out from worse information than it
    /// should have been.
    Degraded,
}

/// A reason this review might be wrong. Dagger's own findings and adapter notes share
/// one list; `impact` is what a reader acts on.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[cfg_attr(feature = "typescript", derive(ts_rs::TS))]
pub struct Warning {
    pub impact: Impact,
    /// Worded here rather than on the page, so there's one wording and names can be
    /// looked up while they're still to hand.
    pub message: String,
    /// What it's about, when it's about one definition. For pointing somebody at it.
    pub about: Option<Identity>,
}

/// One definition, and everything this review knows about it.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[cfg_attr(feature = "typescript", derive(ts_rs::TS))]
pub struct Definition {
    /// Whether it holds others. Said by whoever read the file.
    pub role: Role,
    /// What it was on each side.
    pub sides: Sides,
    /// What happened to it.
    pub change: Change,
    /// Hops from the nearest change that reached it. Never zero: its own change is `change`.
    pub reached: Option<NonZeroU32>,
    /// The one word for what happened to it, as every rendering shows it.
    pub mark: Mark,
    /// What it's written inside. Always present in this review when it isn't `None`.
    pub parent: Option<Identity>,
}

/// Everything one review of a repository comes to.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[cfg_attr(feature = "typescript", derive(ts_rs::TS))]
pub struct Review {
    /// The commit's title, when the adapter knows one. A directory has no commit message.
    pub title: Option<String>,
    pub definitions: BTreeMap<Identity, Definition>,
    /// What to read, in order. Whatever a step names is worth reading; everything else in
    /// `definitions` is here to be drawn around it.
    pub reading: Vec<Step>,
    pub edges: Vec<Edge>,
    /// Everything drawn, as it nests and stacks, so the reading and the page can't disagree.
    pub groups: Vec<Group>,
    /// What a reader calls the grouping: "package", "crate".
    pub grouping: Option<String>,
    /// How far this review was told to follow a change outward.
    pub ripples: u32,
    pub cost: Cost,
    pub warnings: Vec<Warning>,
}

/// `notes` are the adapters' warnings, already worded. `found` is what dagger noticed,
/// worded here where the definitions are still to hand.
pub fn review(
    definitions: Vec<model::Definition>,
    references: &[Reference],
    ripples: u32,
    grouping: &Grouping,
    notes: Vec<Warning>,
    found: Vec<Diagnostic>,
    title: Option<String>,
) -> Review {
    let (classified, lopsided): (Vec<(&model::Definition, Change)>, Vec<Option<Diagnostic>>) =
        definitions
            .iter()
            .map(|def| {
                let (change, found) = classify(def);
                ((def, change), found)
            })
            .unzip();
    let changes: BTreeMap<Identity, Change> = classified
        .iter()
        .map(|(def, change)| (def.identity, change.clone()))
        .collect();

    let (reached, unbound) = reached(&changes, references, ripples);
    let findings: Vec<Diagnostic> = found
        .into_iter()
        .chain(lopsided.into_iter().flatten())
        .chain(unbound)
        .collect();

    let warnings: Vec<Warning> = notes
        .into_iter()
        .chain(findings.iter().map(|one| told(one, &definitions)))
        .collect();

    let read: BTreeSet<Identity> = changes
        .iter()
        .filter(|(_, change)| change.worth_reading())
        .map(|(identity, _)| *identity)
        .chain(reached.keys().copied())
        .collect();

    let latest: BTreeMap<Identity, &model::Occurrence> = definitions
        .iter()
        .map(|def| (def.identity, def.sides.latest()))
        .collect();
    let parent_of: BTreeMap<Identity, Identity> = {
        let by_name: BTreeMap<&Locator, Identity> = latest
            .iter()
            .map(|(identity, it)| (&it.locator, *identity))
            .collect();
        latest
            .iter()
            .filter_map(|(identity, it)| Some((*identity, *by_name.get(it.parent.as_ref()?)?)))
            .collect()
    };

    // Everything worth reading plus every container around it, so a box is never drawn
    // around something whose container isn't here.
    let shown: BTreeSet<Identity> = read
        .iter()
        .flat_map(|&one| std::iter::successors(Some(one), |at| parent_of.get(at).copied()))
        .collect();

    let edges = project(&shown, references);

    let kept: BTreeMap<Identity, Definition> = classified
        .into_iter()
        .filter(|(def, _)| shown.contains(&def.identity))
        .map(|(def, change)| {
            let identity = def.identity;
            (
                identity,
                Definition {
                    role: def.sides.latest().role,
                    mark: change.mark(reached.contains_key(&identity)),
                    change,
                    reached: reached.get(&identity).copied(),
                    parent: parent_of.get(&identity).copied(),
                    sides: def.sides.clone(),
                },
            )
        })
        .collect();

    let shape = Shape::new(&kept, grouping);
    let ordering = order(&read, &edges, &definitions, grouping);
    let groups = shape.groups(&edges, &ordering.steps);

    Review {
        title,
        groups,
        grouping: (!grouping.name.is_empty()).then(|| grouping.name.clone()),
        definitions: kept,
        reading: ordering.steps,
        cost: ordering.cost,
        edges,
        ripples,
        warnings,
    }
}

/// One of dagger's own findings, worded. Names are looked up here so five findings of
/// one kind can be told apart.
fn told(found: &Diagnostic, definitions: &[model::Definition]) -> Warning {
    let named = |identity: Identity| match definitions
        .iter()
        .find(|def| def.identity == identity)
        .map(|def| def.sides.latest())
    {
        Some(shows) => format!("{} ({})", shows.locator, shows.file),
        None => format!("definition {}", identity.0),
    };

    let (impact, message) = match found {
        Diagnostic::Unattributed { file, lines, at } => (
            Impact::Incomplete,
            format!(
                "{file}: {lines} changed line{} belong to no definition, around line {}",
                if *lines == 1 { "" } else { "s" },
                at.iter().map(u32::to_string).collect::<Vec<_>>().join(", ")
            ),
        ),
        Diagnostic::UnboundInType { definition, symbol } => (
            Impact::Incomplete,
            format!(
                "{symbol} appears where callers of {} can see it, but nothing could say \
                 what it refers to",
                named(*definition)
            ),
        ),
        Diagnostic::MentionFromNowhere { from } => (
            Impact::Incomplete,
            format!("a mention came from {from}, which was never reported"),
        ),
        Diagnostic::Tangled { definition, at } => (
            Impact::Incomplete,
            format!(
                "{definition} was handed over with two of its pieces covering the same text, \
                 around byte {at} — so a line of it is shown twice, and read as two different \
                 kinds of change"
            ),
        ),
        Diagnostic::TwoOfOneName { locator, times } => (
            Impact::Incomplete,
            format!(
                "{times} definitions are called {locator}, so only one of them could be \
                 followed from one side to the other"
            ),
        ),
        Diagnostic::LopsidedType { definition } => (
            Impact::Degraded,
            format!(
                "{}: the compiler described one side of this and not the other, so the \
                 signature as written was compared instead",
                named(*definition)
            ),
        ),
    };

    Warning {
        impact,
        message,
        about: found.about(),
    }
}

/// A mention counts if it was there in either snapshot, so that a deleted definition
/// still hangs off whatever it used to call.
fn dependencies(references: &[Reference]) -> BTreeMap<Identity, Vec<Identity>> {
    references
        .iter()
        .filter_map(|reference| match reference.to {
            Target::Known(to) if reference.from != to => Some((reference.from, to)),
            _ => None,
        })
        .fold(BTreeMap::new(), |mut out, (from, to)| {
            out.entry(from).or_default().push(to);
            out
        })
}

/// Walk out from each definition worth drawing, stepping over anything the reader won't
/// see, so two changed definitions joined by an untouched helper still look joined.
fn project(shown: &BTreeSet<Identity>, references: &[Reference]) -> Vec<Edge> {
    let dependencies = dependencies(references);
    let mut edges = Vec::new();

    for from in shown {
        let mut seen = BTreeSet::from([*from]);
        let mut queue: VecDeque<Identity> = dependencies
            .get(from)
            .into_iter()
            .flatten()
            .copied()
            .collect();

        while let Some(node) = queue.pop_front() {
            if !seen.insert(node) {
                continue;
            }
            if shown.contains(&node) {
                edges.push(Edge {
                    from: *from,
                    to: node,
                });
                continue;
            }
            queue.extend(dependencies.get(&node).into_iter().flatten().copied());
        }
    }

    edges
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::model::{Part, Sides};
    use crate::testing::{occurrence, reference};

    fn edited(id: u32, name: &str) -> model::Definition {
        model::Definition {
            identity: Identity(id),
            sides: Sides::Kept {
                before: occurrence(name, &[(Part::Body, "old")]),
                after: occurrence(name, &[(Part::Body, "new")]),
            },
        }
    }

    fn untouched(id: u32, name: &str) -> model::Definition {
        model::Definition {
            identity: Identity(id),
            sides: Sides::Kept {
                before: occurrence(name, &[(Part::Body, "same")]),
                after: occurrence(name, &[(Part::Body, "same")]),
            },
        }
    }

    fn reviewed(
        definitions: Vec<model::Definition>,
        references: &[Reference],
        ripples: u32,
    ) -> Review {
        review(
            definitions,
            references,
            ripples,
            &Grouping::default(),
            Vec::new(),
            Vec::new(),
            None,
        )
    }

    /// Everything handed over is either worth reading or holds something that is.
    #[test]
    fn only_definitions_worth_reading_get_in() {
        let review = reviewed(
            vec![edited(0, "lookupPrice"), untouched(1, "withRetry")],
            &[],
            1,
        );

        assert_eq!(
            review.definitions.keys().copied().collect::<Vec<_>>(),
            vec![Identity(0)]
        );
        assert_eq!(review.reading.len(), 1);
    }

    /// buildLineItems calls withRetry calls lookupPrice. The helper is untouched and
    /// generic, so the reader never sees it, but the two edits are still related.
    #[test]
    fn an_untouched_helper_is_stepped_over() {
        let review = reviewed(
            vec![
                edited(0, "buildLineItems"),
                untouched(1, "withRetry"),
                edited(2, "lookupPrice"),
            ],
            &[reference(0, 1, Part::Body), reference(1, 2, Part::Body)],
            u32::MAX,
        );

        assert_eq!(
            review.edges,
            vec![Edge {
                from: Identity(0),
                to: Identity(2),
            }]
        );
    }

    #[test]
    fn a_direct_mention_is_an_edge() {
        let review = reviewed(
            vec![edited(0, "cartTotal"), edited(1, "lineTotal")],
            &[reference(0, 1, Part::Body)],
            u32::MAX,
        );

        assert_eq!(
            review.edges,
            vec![Edge {
                from: Identity(0),
                to: Identity(1),
            }]
        );
    }

    #[test]
    fn a_dead_end_helper_produces_no_edge() {
        let review = reviewed(
            vec![edited(0, "cartTotal"), untouched(1, "log")],
            &[reference(0, 1, Part::Body)],
            u32::MAX,
        );

        assert!(review.edges.is_empty());
    }

    /// Every identity named anywhere is a definition that was handed over.
    #[test]
    fn nothing_names_a_definition_that_was_not_handed_over() {
        let review = reviewed(
            vec![
                edited(0, "buildLineItems"),
                untouched(1, "withRetry"),
                edited(2, "lookupPrice"),
            ],
            &[reference(0, 1, Part::Body), reference(1, 2, Part::Body)],
            u32::MAX,
        );

        let known = |identity: &Identity| review.definitions.contains_key(identity);
        for edge in &review.edges {
            assert!(
                known(&edge.from) && known(&edge.to),
                "edge {edge:?} dangles"
            );
        }
        for step in &review.reading {
            assert!(known(&step.definition), "a step names nobody");
            assert!(step.on_faith.iter().all(known), "on_faith names nobody");
        }
        for (identity, one) in &review.definitions {
            if let Some(parent) = one.parent {
                assert!(known(&parent), "{identity:?} is inside nobody");
                assert_eq!(
                    review.definitions[&parent].role,
                    Role::Container,
                    "{identity:?} is inside something that holds nothing"
                );
            }
        }
        for warning in &review.warnings {
            assert!(warning.about.is_none_or(|about| known(&about)));
            assert!(!warning.message.is_empty(), "a warning nobody can read");
        }
    }

    /// Nothing is written inside itself, however the extractors describe it.
    #[test]
    fn containment_does_not_go_in_circles() {
        let review = reviewed(vec![edited(0, "one"), edited(1, "two")], &[], 1);

        for identity in review.definitions.keys() {
            let mut seen = BTreeSet::from([*identity]);
            let mut at = review.definitions[identity].parent;
            while let Some(one) = at {
                assert!(seen.insert(one), "{identity:?} is inside itself");
                at = review.definitions[&one].parent;
            }
        }
    }
}
